File: frontend/src/lib/motor-normalizer.ts
/**
 * Espelha `utils/motor_normalizer.py` para lista/detalhe Next.js (Supabase direto).
 */

export type MotorUiFields = Record<string, string>;

const A_PASSO_P = ["PassoPrincipal", "passo_principal", "PassosPrincipais", "passos_principais"] as const;
const A_PASSO_A = ["PassoAuxiliar", "passo_auxiliar", "PassosAuxiliares", "passos_auxiliares"] as const;
const A_ESP_P = ["EspirasPrincipal", "espiras_principal", "EspirasPrincipais", "espiras_principais"] as const;
const A_ESP_A = ["EspirasAuxiliar", "espiras_auxiliar", "EspirasAuxiliares", "espiras_auxiliares"] as const;
const A_FIO_P = ["FioPrincipal", "fio_principal", "FiosPrincipais", "fios_principais"] as const;
const A_FIO_A = ["FioAuxiliar", "fio_auxiliar", "FiosAuxiliares", "fios_auxiliares"] as const;
const A_LIG_P = ["LigacaoPrincipal", "ligacao_principal", "LigacoesPrincipais", "ligacoes_principais"] as const;
const A_LIG_A = ["LigacaoAuxiliar", "ligacao_auxiliar", "LigacoesAuxiliares", "ligacoes_auxiliares"] as const;
const A_EIXO_X = ["EixoX", "eixo_x"] as const;
const A_EIXO_Y = ["EixoY", "eixo_y"] as const;
const A_EIXO = ["Eixo", "eixo"] as const;
const A_MEDIDAS = ["Medidas", "medidas"] as const;
const A_POT = ["Potencia", "potencia"] as const;
const A_RPM = ["Rpm", "rpm", "RPM"] as const;
const A_TENSAO = ["Tensao", "tensao"] as const;
const A_CORR = ["Corrente", "corrente"] as const;
const A_FREQ = ["Frequencia", "frequencia"] as const;
const A_POLOS = ["Polos", "polos"] as const;
const A_TIPO = ["TipoMotor", "tipo_motor"] as const;
const A_CARCACA = ["Carcaca", "carcaca"] as const;
const A_PASSO_MULTI = ["Passo", "passo", "Passos", "passos"] as const;
const A_ESP_MULTI = ["Espiras", "espiras"] as const;
const A_FIO_MULTI = ["Fio", "fio", "Fios", "fios"] as const;
const A_LIG_MULTI = ["Ligacao", "ligacao", "Ligacoes", "ligacoes"] as const;

export function isEmpty(value: unknown): boolean {
  if (value === null || value === undefined) return true;
  if (typeof value === "string") {
    const t = value.trim();
    if (!t) return true;
    const low = t.toLowerCase();
    if (low === "-" || low === "null" || low === "none" || low === "nan") return true;
  }
  if (Array.isArray(value) && value.length === 0) return true;
  if (typeof value === "object" && !Array.isArray(value) && Object.keys(value as object).length === 0) return true;
  return false;
}

function variaveisSite(row: Record<string, unknown>): Record<string, unknown> {
  let raw = row.VariaveisSite ?? row.variaveis_site ?? row.Variaveis_site;
  if (raw && typeof raw === "object" && !Array.isArray(raw)) return raw as Record<string, unknown>;
  if (typeof raw === "string" && raw.trim().startsWith("{")) {
    try {
      const p = JSON.parse(raw) as unknown;
      return p && typeof p === "object" && !Array.isArray(p) ? (p as Record<string, unknown>) : {};
    } catch {
      return {};
    }
  }
  return {};
}

export function pickValue(row: Record<string, unknown>, aliases: readonly string[]): unknown {
  const vs = variaveisSite(row);
  for (const alias of aliases) {
    const v = row[alias];
    if (!isEmpty(v)) return v;
    if (Object.prototype.hasOwnProperty.call(vs, alias)) {
      const v2 = vs[alias];
      if (!isEmpty(v2)) return v2;
    }
  }
  return undefined;
}

export function asDisplayText(value: unknown): string {
  if (isEmpty(value)) return "";
  if (Array.isArray(value)) {
    const parts = value.map((x) => String(x).trim()).filter((x) => !isEmpty(x));
    return parts.join(", ");
  }
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const o = value as Record<string, unknown>;
    const parts = Object.values(o)
      .map((v) => String(v).trim())
      .filter((v) => !isEmpty(v));
    return parts.join(", ");
  }
  return String(value).trim();
}

function itemsFromMultivalue(raw: unknown): string[] {
  if (isEmpty(raw)) return [];
  if (Array.isArray(raw)) {
    const out: string[] = [];
    for (const x of raw) {
      const s = asDisplayText(x);
      if (s) out.push(s);
    }
    return out;
  }
  if (typeof raw === "string") {
    const t = raw.trim();
    if (!t) return [];
    const parts = t.split(/\s*[,;]\s*/).filter((p) => p.trim());
    return parts.length > 1 ? parts : [t];
  }
  const s = asDisplayText(raw);
  return s ? [s] : [];
}

export function normalizeMotorRowForUi(row: Record<string, unknown> | null | undefined): MotorUiFields {
  const r = row || {};

  let passoPrincipal = asDisplayText(pickValue(r, A_PASSO_P));
  let passoAuxiliar = asDisplayText(pickValue(r, A_PASSO_A));
  let espirasPrincipal = asDisplayText(pickValue(r, A_ESP_P));
  let espirasAuxiliar = asDisplayText(pickValue(r, A_ESP_A));
  let fioPrincipal = asDisplayText(pickValue(r, A_FIO_P));
  let fioAuxiliar = asDisplayText(pickValue(r, A_FIO_A));
  let ligacaoPrincipal = asDisplayText(pickValue(r, A_LIG_P));
  let ligacaoAuxiliar = asDisplayText(pickValue(r, A_LIG_A));

  const eixoX = asDisplayText(pickValue(r, A_EIXO_X));
  const eixoY = asDisplayText(pickValue(r, A_EIXO_Y));
  let eixo = asDisplayText(pickValue(r, A_EIXO));
  let medidas = asDisplayText(pickValue(r, A_MEDIDAS));

  const potencia = asDisplayText(pickValue(r, A_POT));
  const rpm = asDisplayText(pickValue(r, A_RPM));
  const tensao = asDisplayText(pickValue(r, A_TENSAO));
  const corrente = asDisplayText(pickValue(r, A_CORR));
  const frequencia = asDisplayText(pickValue(r, A_FREQ));
  const polos = asDisplayText(pickValue(r, A_POLOS));
  const tipoMotor = asDisplayText(pickValue(r, A_TIPO));
  const carcaca = asDisplayText(pickValue(r, A_CARCACA));

  const passoItems = itemsFromMultivalue(pickValue(r, A_PASSO_MULTI));
  if (!passoPrincipal && passoItems.length) passoPrincipal = passoItems[0]!;
  if (!passoAuxiliar && passoItems.length >= 2) passoAuxiliar = passoItems[1]!;

  const espItems = itemsFromMultivalue(pickValue(r, A_ESP_MULTI));
  if (!espirasPrincipal && espItems.length) espirasPrincipal = espItems[0]!;
  if (!espirasAuxiliar && espItems.length >= 2) espirasAuxiliar = espItems[1]!;

  const fioItems = itemsFromMultivalue(pickValue(r, A_FIO_MULTI));
  if (!fioPrincipal && fioItems.length) fioPrincipal = fioItems[0]!;
  if (!fioAuxiliar && fioItems.length >= 2) fioAuxiliar = fioItems[1]!;

  const ligItems = itemsFromMultivalue(pickValue(r, A_LIG_MULTI));
  if (!ligacaoPrincipal && ligItems.length) ligacaoPrincipal = ligItems[0]!;
  if (!ligacaoAuxiliar && ligItems.length >= 2) ligacaoAuxiliar = ligItems[1]!;

  if (!eixo && (eixoX || eixoY)) {
    eixo = `X:${eixoX || "-"} | Y:${eixoY || "-"}`;
  }
  if (!medidas && (eixoX || eixoY)) {
    medidas = eixoX && eixoY ? `${eixoX} x ${eixoY}` : eixoX || eixoY;
  }

  return {
    passo_principal: passoPrincipal,
    passo_auxiliar: passoAuxiliar,
    espiras_principal: espirasPrincipal,
    espiras_auxiliar: espirasAuxiliar,
    fio_principal: fioPrincipal,
    fio_auxiliar: fioAuxiliar,
    ligacao_principal: ligacaoPrincipal,
    ligacao_auxiliar: ligacaoAuxiliar,
    eixo_x: eixoX,
    eixo_y: eixoY,
    eixo,
    medidas,
    potencia,
    rpm,
    tensao,
    corrente,
    frequencia,
    polos,
    tipo_motor: tipoMotor,
    carcaca,
  };
}

File: utils/motor_normalizer.py
"""
Normalização única de campos de motor vindos do Supabase (view/tabela + VariaveisSite).
Destinado ao Streamlit e a futura API/Next.js — sem efeitos colaterais de UI.
"""

from __future__ import annotations

import json
import re
from typing import Any, Dict, List, Sequence

ALIASES_PASSO_PRINCIPAL = (
    "PassoPrincipal",
    "passo_principal",
    "PassosPrincipais",
    "passos_principais",
)
ALIASES_PASSO_AUXILIAR = (
    "PassoAuxiliar",
    "passo_auxiliar",
    "PassosAuxiliares",
    "passos_auxiliares",
)
ALIASES_ESPIRAS_PRINCIPAL = (
    "EspirasPrincipal",
    "espiras_principal",
    "EspirasPrincipais",
    "espiras_principais",
)
ALIASES_ESPIRAS_AUXILIAR = (
    "EspirasAuxiliar",
    "espiras_auxiliar",
    "EspirasAuxiliares",
    "espiras_auxiliares",
)
ALIASES_FIO_PRINCIPAL = (
    "FioPrincipal",
    "fio_principal",
    "FiosPrincipais",
    "fios_principais",
)
ALIASES_FIO_AUXILIAR = (
    "FioAuxiliar",
    "fio_auxiliar",
    "FiosAuxiliares",
    "fios_auxiliares",
)
ALIASES_LIGACAO_PRINCIPAL = (
    "LigacaoPrincipal",
    "ligacao_principal",
    "LigacoesPrincipais",
    "ligacoes_principais",
)
ALIASES_LIGACAO_AUXILIAR = (
    "LigacaoAuxiliar",
    "ligacao_auxiliar",
    "LigacoesAuxiliares",
    "ligacoes_auxiliares",
)
ALIASES_EIXO_X = ("EixoX", "eixo_x")
ALIASES_EIXO_Y = ("EixoY", "eixo_y")
ALIASES_EIXO = ("Eixo", "eixo")
ALIASES_MEDIDAS = ("Medidas", "medidas")
ALIASES_POTENCIA = ("Potencia", "potencia")
ALIASES_RPM = ("Rpm", "rpm")
ALIASES_TENSAO = ("Tensao", "tensao")
ALIASES_CORRENTE = ("Corrente", "corrente")
ALIASES_FREQUENCIA = ("Frequencia", "frequencia")
ALIASES_POLOS = ("Polos", "polos")
ALIASES_TIPO_MOTOR = ("TipoMotor", "tipo_motor")
ALIASES_CARCACA = ("Carcaca", "carcaca")

ALIASES_PASSO_MULTI = ("Passo", "passo", "Passos", "passos")
ALIASES_ESPIRAS_MULTI = ("Espiras", "espiras")
ALIASES_FIO_MULTI = ("Fio", "fio", "Fios", "fios")
ALIASES_LIGACAO_MULTI = ("Ligacao", "ligacao", "Ligacoes", "ligacoes")


def is_empty(value: Any) -> bool:
    if value is None:
        return True
    if isinstance(value, str):
        t = value.strip()
        if not t:
            return True
        low = t.lower()
        if low in ("-", "null", "none", "nan"):
            return True
    if value == [] or value == {}:
        return True
    return False


def _variaveis_site(row: Dict[str, Any]) -> Dict[str, Any]:
    raw = row.get("VariaveisSite")
    if raw is None:
        raw = row.get("variaveis_site") or row.get("Variaveis_site")
    if isinstance(raw, dict):
        return raw
    if isinstance(raw, str) and raw.strip().startswith("{"):
        try:
            parsed = json.loads(raw)
            return parsed if isinstance(parsed, dict) else {}
        except Exception:
            return {}
    return {}


def pick_value(row: Dict[str, Any], aliases: Sequence[str]) -> Any:
    """
    Para cada alias na ordem:
    1) row[alias] se não vazio
    2) VariaveisSite[alias] se existir e não vazio
    """
    vs = _variaveis_site(row)
    for alias in aliases:
        v = row.get(alias)
        if not is_empty(v):
            return v
        if isinstance(vs, dict) and alias in vs:
            v2 = vs.get(alias)
            if not is_empty(v2):
                return v2
    return None


def as_display_text(value: Any) -> str:
    """Texto para UI; listas viram ', '; vazios viram ''."""
    if is_empty(value):
        return ""
    if isinstance(value, list):
        parts = [str(x).strip() for x in value if not is_empty(x)]
        return ", ".join(parts)
    if isinstance(value, dict):
        if not value:
            return ""
        parts = [str(v).strip() for v in value.values() if not is_empty(v)]
        return ", ".join(parts)
    return str(value).strip()


def _items_from_multivalue(raw: Any) -> List[str]:
    """Extrai lista ordenada de itens (array, string CSV/SC)."""
    if is_empty(raw):
        return []
    if isinstance(raw, list):
        out: List[str] = []
        for x in raw:
            s = as_display_text(x)
            if s:
                out.append(s)
        return out
    if isinstance(raw, str):
        t = raw.strip()
        if not t:
            return []
        parts = [p.strip() for p in re.split(r"\s*[,;]\s*", t) if p.strip()]
        return parts if len(parts) > 1 else [t]
    s = as_display_text(raw)
    return [s] if s else []


def normalize_motor_row_for_ui(row: Dict[str, Any]) -> Dict[str, str]:
    """
    Camada única de normalização para exibição / JSON estável (ex.: Next).
    Todas as chaves em snake_case; valores sempre str (vazio = dado ausente).
    """
    r = row or {}

    passo_principal = as_display_text(pick_value(r, ALIASES_PASSO_PRINCIPAL))
    passo_auxiliar = as_display_text(pick_value(r, ALIASES_PASSO_AUXILIAR))
    espiras_principal = as_display_text(pick_value(r, ALIASES_ESPIRAS_PRINCIPAL))
    espiras_auxiliar = as_display_text(pick_value(r, ALIASES_ESPIRAS_AUXILIAR))
    fio_principal = as_display_text(pick_value(r, ALIASES_FIO_PRINCIPAL))
    fio_auxiliar = as_display_text(pick_value(r, ALIASES_FIO_AUXILIAR))
    ligacao_principal = as_display_text(pick_value(r, ALIASES_LIGACAO_PRINCIPAL))
    ligacao_auxiliar = as_display_text(pick_value(r, ALIASES_LIGACAO_AUXILIAR))

    eixo_x = as_display_text(pick_value(r, ALIASES_EIXO_X))
    eixo_y = as_display_text(pick_value(r, ALIASES_EIXO_Y))
    eixo = as_display_text(pick_value(r, ALIASES_EIXO))
    medidas = as_display_text(pick_value(r, ALIASES_MEDIDAS))

    potencia = as_display_text(pick_value(r, ALIASES_POTENCIA))
    rpm = as_display_text(pick_value(r, ALIASES_RPM))
    tensao = as_display_text(pick_value(r, ALIASES_TENSAO))
    corrente = as_display_text(pick_value(r, ALIASES_CORRENTE))
    frequencia = as_display_text(pick_value(r, ALIASES_FREQUENCIA))
    polos = as_display_text(pick_value(r, ALIASES_POLOS))
    tipo_motor = as_display_text(pick_value(r, ALIASES_TIPO_MOTOR))
    carcaca = as_display_text(pick_value(r, ALIASES_CARCACA))

    # Fallbacks inteligentes (Passo / Espiras / Fio / Ligacao)
    passo_items = _items_from_multivalue(pick_value(r, ALIASES_PASSO_MULTI))
    if not passo_principal and passo_items:
        passo_principal = passo_items[0]
    if not passo_auxiliar and len(passo_items) >= 2:
        passo_auxiliar = passo_items[1]

    esp_items = _items_from_multivalue(pick_value(r, ALIASES_ESPIRAS_MULTI))
    if not espiras_principal and esp_items:
        espiras_principal = esp_items[0]
    if not espiras_auxiliar and len(esp_items) >= 2:
        espiras_auxiliar = esp_items[1]

    fio_items = _items_from_multivalue(pick_value(r, ALIASES_FIO_MULTI))
    if not fio_principal and fio_items:
        fio_principal = fio_items[0]
    if not fio_auxiliar and len(fio_items) >= 2:
        fio_auxiliar = fio_items[1]

    lig_items = _items_from_multivalue(pick_value(r, ALIASES_LIGACAO_MULTI))
    if not ligacao_principal and lig_items:
        ligacao_principal = lig_items[0]
    if not ligacao_auxiliar and len(lig_items) >= 2:
        ligacao_auxiliar = lig_items[1]

    if not eixo and (eixo_x or eixo_y):
        eixo = f"X:{eixo_x or '-'} | Y:{eixo_y or '-'}"

    if not medidas and (eixo_x or eixo_y):
        if eixo_x and eixo_y:
            medidas = f"{eixo_x} x {eixo_y}"
        else:
            medidas = eixo_x or eixo_y

    return {
        "passo_principal": passo_principal,
        "passo_auxiliar": passo_auxiliar,
        "espiras_principal": espiras_principal,
        "espiras_auxiliar": espiras_auxiliar,
        "fio_principal": fio_principal,
        "fio_auxiliar": fio_auxiliar,
        "ligacao_principal": ligacao_principal,
        "ligacao_auxiliar": ligacao_auxiliar,
        "eixo_x": eixo_x,
        "eixo_y": eixo_y,
        "eixo": eixo,
        "medidas": medidas,
        "potencia": potencia,
        "rpm": rpm,
        "tensao": tensao,
        "corrente": corrente,
        "frequencia": frequencia,
        "polos": polos,
        "tipo_motor": tipo_motor,
        "carcaca": carcaca,
    }
